### PR TITLE
PL-5942 migrate metrics collector from Dropwizard to Prometheus

### PR DESCRIPTION
**Pull request by mr-git**

### CODE REVIEW

File: src/main/scala/com/evolutiongaming/scaffeine/MetricsStatsCounter.scala [hunk 1]
```diff
@@ -1,44 +1,57 @@
...
+class MetricsStatsCounter(registry: CollectorRegistry, prefix: String) extends StatsCounter {
 
-  private val misses = metrics.counter("misses")
+  private def createCounter(name: String): Counter = {
+    val c = Counter.build().name(s"${prefix}_$name".replaceAll("\\.", "_")).help(s"$prefix $name").create()
+    registry.register(c)
+    c
+  }
+
+//  private def createHistogram(name: String): Histogram = {
```
Comment (t3hnar) [inline src/main/scala/com/evolutiongaming/scaffeine/MetricsStatsCounter.scala]: will you remove this?

File: src/main/scala/com/evolutiongaming/scaffeine/MetricsStatsCounter.scala [hunk 2]
```diff
@@ -1,44 +1,57 @@
...
 
-  override def recordHits(i: Int): Unit = hits += i.toLong
+  override def recordHits(i: Int): Unit = hits.inc(i.toDouble)
 
   override def recordEviction(): Unit = eviction.inc()
 
   override def recordLoadFailure(l: Long): Unit = {
     failure.inc()
-    time.update(Duration.fromNanos(l))
+//    time.observe(Duration.fromNanos(l).toMillis)
```
Comment (t3hnar) [inline src/main/scala/com/evolutiongaming/scaffeine/MetricsStatsCounter.scala]: ?

File: src/main/scala/com/evolutiongaming/scaffeine/MetricsStatsCounter.scala [hunk 3]
```diff
@@ -1,44 +1,57 @@
...
+    time.inc(Duration.fromNanos(l).toMillis.toDouble)
   }
 
   override def recordLoadSuccess(l: Long): Unit = {
     success.inc()
-    time.update(Duration.fromNanos(l))
+//    time.observe(Duration.fromNanos(l).toMillis)
```
Comment (t3hnar) [inline src/main/scala/com/evolutiongaming/scaffeine/MetricsStatsCounter.scala]: ?

**Review by t3hnar — COMMENTED** [review]
please remove some commented out code.

File: src/main/scala/com/evolutiongaming/scaffeine/MetricsStatsCounter.scala [hunk 4]
```diff
@@ -1,44 +1,57 @@
...
+  private val eviction = createCounter("evictions")
 
-  private val failure = metrics.counter("load_failure")
+  private val failure = createCounter("load_failure")
 
-  private val success = metrics.counter("load_success")
+  private val success = createCounter("load_success")
 
-  private val time = metrics.timer("load_time")
+//  private val time = createHistogram("load_time")
```
Comment (stavanets) [inline src/main/scala/com/evolutiongaming/scaffeine/MetricsStatsCounter.scala]: could you please remove this?

File: src/main/scala/com/evolutiongaming/scaffeine/MetricsStatsCounter.scala (hunk 1 above)
Comment (mr-git) [reply src/main/scala/com/evolutiongaming/scaffeine/MetricsStatsCounter.scala]: sure

File: src/main/scala/com/evolutiongaming/scaffeine/MetricsStatsCounter.scala (hunk 2 above)
Comment (mr-git) [reply src/main/scala/com/evolutiongaming/scaffeine/MetricsStatsCounter.scala]: will be removed, as `Histogram` is not needed

File: src/main/scala/com/evolutiongaming/scaffeine/MetricsStatsCounter.scala (hunk 3 above)
Comment (mr-git) [reply src/main/scala/com/evolutiongaming/scaffeine/MetricsStatsCounter.scala]: will be removed, as `Histogram` is not needed

File: src/main/scala/com/evolutiongaming/scaffeine/MetricsStatsCounter.scala (hunk 4 above)
Comment (mr-git) [reply src/main/scala/com/evolutiongaming/scaffeine/MetricsStatsCounter.scala]: sure, I will cleanup before/during merge

**Comment by mr-git** [comment]
cleaned and squashed

**Comment by t3hnar** [comment]
LGTM

File: src/main/scala/com/evolutiongaming/scaffeine/MetricsStatsCounter.scala [hunk 5]
```diff
@@ -1,44 +1,48 @@
...
+  private val eviction = createCounter("evictions")
 
-  private val failure = metrics.counter("load_failure")
+  private val failure = createCounter("load_failure")
 
-  private val success = metrics.counter("load_success")
+  private val success = createCounter("load_success")
 
-  private val time = metrics.timer("load_time")
+  private val time = createCounter("load_time")
```
Comment (hulevskyi) [inline src/main/scala/com/evolutiongaming/scaffeine/MetricsStatsCounter.scala:26]: Is it ok that we will have counter instead of timer for `load_time`?

Comment (mr-git) [reply src/main/scala/com/evolutiongaming/scaffeine/MetricsStatsCounter.scala:26]: Here `Timer` directly was used just for summing durations.
Prometheus doesn't provide similar structure, except `Histogram.Timer`, which requires start/stop calls, but we get only durations here. Also, `Histogram` doesn't provide means to get a `count`, which we need here for statistics.
Only thing that is missing now are quantiles, though I do not see it as a big problem.

**Pull request merged**